New Page: aria-live (#10633)
mdn/content: files/en-us/web/accessibility/aria/attributes/aria-live/index.md

---
title: 'aria-live'
slug: Web/Accessibility/ARIA/Attributes/aria-live
tags: 
  - Accessibility
  - ARIA
  - ARIA attribute
  - ARIA property
  - aria-live
  - Reference
---

The global `aria-live` attribute indicates that an element will be updated, and describes the types of updates the user agents, assistive technologies, and user can expect from the live region.

## Description

> Sighted users will generally notice live updates, such as live sports scores, news, and stock market tickers, when they appear on screen, if they're styled in a way to be noticeable. Updating areas of content are called "live regions". Sometimes the live region information is important for a user to know immediately. Sometimes updated information, like stock prices, are background noise to those reading a finance article on a page with stock ticker. As assitive technology users can't "see" these live updated, the `aria-live` attribute is used to define what updated information the user should be made aware of immediately, what they should be informed about if the opportunity avails itself, and what updated they don't need to be proactively informed about but, rather, can learn when they choose to focus on the updated area.  

Reviewer: ```suggestion
If they're styled in a way to be noticeable, most sighted users will generally notice live updates, such as live sports scores, news, and stock market tickers when they appear on screen.

Updating areas of content such as these are called "live regions". Sometimes the live region information is important for a user to know immediately. Sometimes updated information, like stock prices, are background noise to those reading a finance article on a page with stock ticker. 

As some assistive technology users can't "see" these live updates, the `aria-live` attribute is used to define what updated information the user should be:

- Made aware of immediately, 
- Informed about if the opportunity avails itself, and 
- Proactively informed about but can learn when they choose to focus on the updated area. 
```

Updating to communicate not all AT users are blind or low vision, breaking out the list of update types for easier scanning.

## Description

When content changes after initial load, assistive technology (AT) users may not "see" the changes. Some changes are important. Others are not. The `aria-live` attribute enables developers to inform user of updates and choose, based on importance and urgency, whether to immediately, proactively, or passively inform AT users of changes to the content. 

If a section of the screen updates, if it's styled in a way to be noticeable, most sighted users will generally notice live updates. Screen readers, on the other hand, only focus on one part of the page at a time; and that part may not be where the update is. The `aria-live` attribute provides a way for developers to annouce such changes to AT based on event triggers set by the developer rather than by user intiated actions so they are made aware that the content has changed. 

```html
<div id="announce" aria-live="polite"></div>
```

The `aria-live` attribute is set on an **empty** element. When an update to the page occures, the empty element with that `aria-live` attribute should be updated with a brief announcement informing the user an update has been made. 

```html

> <div id="announce" aria-live="polite">This message is announced</div>

Reviewer: ```suggestion
<div id="announce" aria-live="polite">
  <p>This message is announced.</p>
</div>
```

```

> When the accessibility API recognizes a change to the live region above, it will announce the contents of that live region based on the value of the attribute.  The element is **not** given focus. 

Reviewer: ```suggestion
When the accessibility API recognizes a change to the live region above, it will announce the contents of that live region based on the value of the attribute. The element is **not** given focus. 
```

> If you want the contents that were updated to be read, used the [`aria-atomic`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-atomic).  Use [`aria-relevant`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-relevant) to define which sections of the update need to be reread to the user. Use [`aria-busy`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-busy) to prevent announcemets while updates are still being made. 

Reviewer: ```suggestion
If you want all the contents of the live region to be read, used the [`aria-atomic`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-atomic). Use [`aria-relevant`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-relevant) to define only the sections of an update need to be reread to the user. Use [`aria-busy`](/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-busy) to prevent announcements while updates are still being made. 
```

### Choosing the `aria-live` value

> As some assitive technology users can't "see" live updates, the `aria-live` attribute is used to define what updated information the user should be:

Reviewer: ```suggestion
As some assitive technology users can't "see" live updates, the `aria-live` attribute is used to define what updated information the user should be:

```

- Made aware of immediately, 
- Informed about if the opportunity avails itself, and 
- Proactively informed about but can learn when they choose to focus on the updated area.

> The value, expressed in degrees of importance, describes the types of updates the user agents, assistive technologies, and user can expect from the live region. When set to `polite`, assistive technologies will notify users of updates but generally do not interrupt the current task, with the updates having a low priority. When set to `assertive`, assistive technologies immediately notify the user, potentially clearing the speech queue of previous updates. 

Reviewer: ```suggestion
`aria-live`'s value describes the types of updates the user agents, assistive technologies, and user can expect from the live region, and is used to express degrees of importance. 

When `aria-live`'s attribute is set to `polite`, assistive technologies will notify users of updates but generally do not interrupt the current task, with the updates having a low priority. When set to `assertive`, assistive technologies immediately notify the user, potentially clearing the speech queue of previous updates. 
```

> Screen readers buffer content when the page is loaded so content added after the initial accessibility tree is built may not be noticed as AT users may start consuming content before dynamic widgets are populated; they may not see the tell tale sign of content moving around as widgets finish loading. You likely want users to know about the page has been updated when the opportunity arises by setting `aria-live="polite"`. 

Reviewer: ```suggestion
Screen readers buffer content when the page is loaded. Because of this, content added after the initial accessibility tree is built may not be noticed as AT users start consuming content before dynamic widgets are populated—users may not know about page or view content updating as widgets finish loading. In this situation you can let users know the page has been updated by setting `aria-live="polite"`. 
```

> Fully populated pages may have updates too. Some pages have real-time sports scores, news crawlers, and stock market tickers. With features that update often like these, unless they are the main function of the page, you likely do want inform the user every time it updates, but do want to inform them the widget does get updated. Here you would set `aria-live="off"`. In these scenarios, there is no reason to inform the user of updates unless the user is focused on the live region. In fact, when [`role="marquee"`](/en-US/docs/Web/Accessibility/ARIA/Roles/marquee_role) is set, the implied value is `aria-live="off"`,

Reviewer: ```suggestion
Fully populated pages may have updates too. Examples are content like real-time sports scores, news crawlers, and stock market tickers. Unless these kinds of updates are the main function of the page, you likely do want inform the user every time it updates, but do want to inform them the widget does get updated. Here, you would set `aria-live="off"`. In these scenarios there is no reason to inform the user of updates unless they are focused on the live region.
```

> Some live updates are important and time sensitive. If you are selling concert tickets and the user has a limited time to make the purchase, you don't want to wait until a lull in activity to tell them their time is almost up (or already over). When being informed as soon as possible is necessary, set `aria-live="assertive"`. If the updated information is an [`alert`](/en-US/docs/Web/Accessibility/ARIA/Roles/alert_role), the `aria-live` attribute is not needed. 

Reviewer: ```suggestion
Some live updates are important and time sensitive. For example, if you are selling concert tickets and the user has a limited time to make the purchase, you don't want to wait until a lull in activity to tell them their time is almost up (or already over). When being informed as soon as possible is necessary, set `aria-live="assertive"`. If the updated information is an [`alert`](/en-US/docs/Web/Accessibility/ARIA/Roles/alert_role), the `aria-live` attribute is not needed.
```

> In this time-limit scenario, to be accessible, you need to also provide a way for users to extend the time available or turn the timer off completely.

Reviewer: ```suggestion
In this time-limit scenario, to be accessible, you need to also [provide a way for users to extend the time available or turn the timer off completely](https://www.w3.org/WAI/WCAG21/Understanding/timing-adjustable.html).
```

### Using `aria-live`

> A live region includes the element and all it's descendants. When not set on updating content, the value comes from the nearest ancestor with a valid `aria-live` attribute value set. When set to `off`, or when the attribute is omitted altogether on the updated element and all the ancestor nodes in the DOM tree, the user is not informed, but will see the updates if they navigate to the live region.

Reviewer: ```suggestion
A live region includes the element and all it's descendants. When not set on updating content, `aria-live`'s value comes from the nearest ancestor with a valid `aria-live` attribute value set. When set to `off`, or when the attribute is omitted altogether on the updated element and all the ancestor nodes in the DOM tree, the user is not informe. Users will still be able to hear the updates if they navigate to the live region, however.
```
